### PR TITLE
feat(restore-indices): add multithreading and add aspectName, urn filter

### PR DESCRIPTION
**Pull request by anshbansal**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java [hunk 1]
```diff
@@ -126,7 +131,7 @@ public Function<UpgradeContext, UpgradeStepResult> executable() {
 
           totalRowsMigrated++;
         }
-        context.report().addLine(String.format("Successfully sent MAEs for %s rows", totalRowsMigrated));
+        context.report().addLine(String.format("Successfully sent MAEs for %s/%s rows (%.2f%%). %s rows ignored (%.2f%%)", totalRowsMigrated, rowCount, (float)totalRowsMigrated*100/rowCount, ignored, (float)ignored*100/rowCount));
```
Comment (shirshanka) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:134]: ensure division by 0 doesn't happen

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java [hunk 2]
```diff
@@ -135,7 +140,7 @@ public Function<UpgradeContext, UpgradeStepResult> executable() {
         }
       }
       if (totalRowsMigrated != rowCount) {
-        context.report().addLine(String.format("Failed to send MAEs for %d rows...", rowCount - totalRowsMigrated));
+        context.report().addLine(String.format("Failed to send MAEs for %d rows (%.2f%%).", rowCount - totalRowsMigrated, (float)(rowCount - totalRowsMigrated)*100/rowCount));
```
Comment (shirshanka) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:143]: ensure division by 0 doesn't happen

**Review by jjoyce0510 — COMMENTED** [review]
Nice improvements! 

I'm a bit nervous about correctness given the scope of changes - have you been able to validate the behavior of this job locally?

Going to also ask that @RyanHolstien takes a pass at this since he's been getting super familiar with this job recently :)

**Comment by anshbansal** [comment]
Smoke test failure has nothing to do with this change. It seems that is broken in latest master.

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java [hunk 3]
```diff
@@ -111,42 +122,185 @@ public Function<UpgradeContext, UpgradeStepResult> executable() {
...
+  public Function<UpgradeContext, UpgradeStepResult> executable() {
+    return (context) -> {
+
+      long startTime = System.currentTimeMillis();
+
+      int batchSize = getBatchSize(context.parsedArgs());
+      int numThreads = getThreadCount(context.parsedArgs());
+      long batchDelayMs = getBatchDelayMs(context.parsedArgs());
+      Optional<String> aspectName;
+      if (containsKey(context.parsedArgs(), RestoreIndices.ASPECT_NAME_ARG_NAME)) {
```
Comment (RyanHolstien) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:202]: I think we should move away from the command line arguments approach, it's less flexible than just pure ENV and requires us to modify the helm chart templates every time we add new configuration.

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java [hunk 4]
```diff
@@ -111,42 +122,185 @@ public Function<UpgradeContext, UpgradeStepResult> executable() {
...
+        if (tmpResult != null) {
+          totalRowsMigrated += tmpResult.rowsMigrated;
+          ignored += tmpResult.ignored;
+          reportStats(context, totalRowsMigrated, ignored, rowCount, startTime);
+        }
+      }
+      executor.shutdown();
       if (totalRowsMigrated != rowCount) {
-        context.report().addLine(String.format("Failed to send MAEs for %d rows...", rowCount - totalRowsMigrated));
+        float percentFailed = 0.0f;
```
Comment (RyanHolstien) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:263]: I feel like raw count of failed rows is usually going to be more useful than a percent.

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java [hunk 5]
```diff
@@ -111,42 +122,185 @@ public Function<UpgradeContext, UpgradeStepResult> executable() {
...
+        if (rowCount > 0) {
+          percentFailed = (float) (rowCount - totalRowsMigrated) * 100 / rowCount;
+        }
+        context.report().addLine(String.format(
+                "Failed to send MAEs for %d rows (%.2f%%).", rowCount - totalRowsMigrated, percentFailed));
```
Comment (RyanHolstien) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:268]: I think the percent should be clarified here, generally I think people would expect to see the percent completed not percent failed so something like

Failed to send MAEs for %d rows, (%.2f%%) of total failed

**Review by RyanHolstien — APPROVED** [review]
No blockers, just comments :)

**Pull request merged**